### PR TITLE
Gazelle: read go_prefix from top-level BUILD if unset and available.  Merge BUILD content with existing

### PR DESCRIPTION
**Pull request by pmbethe09**

### CODE REVIEW

**Comment by pmbethe09** [comment]
@yugui any thoughts?  Did I accidentally re-do the same work you had in progress?

**Comment by yugui** [comment]
@pmbethe09 
np. Although this PR is similar to feature/reconcile branch, I like your implementation more.

File: go/tools/gazelle/merger/merger.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+	}
+	sort.Strings(keys)
+	dest.List = dest.List[:1]
+	for _, k := range keys {
+		dest.List = append(dest.List, vals[k])
+	}
+}
+
+func ruleUsed(rule string, oldfile *bzl.File) bool {
+	for _, s := range oldfile.Stmt {
```
Comment (yugui) [inline go/tools/gazelle/merger/merger.go]: `len(oldfile.Rules(rule)) != 0` might be simpler, although it is less efficient.

File: go/tools/gazelle/merger/merger.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+	load string
+}
+
+func (m *loadMatcher) match(c *bzl.CallExpr) bool {
+	return literal(c.X) == "load" && len(c.List) > 0 && m.load == stringValue(c.List[0])
+}
+
+// get returns the value for the 'name = value' entry in the calls arguments
+// or "" if not found.
+func get(name string, c *bzl.CallExpr) string {
```
Comment (yugui) [inline go/tools/gazelle/merger/merger.go]: These accessor functions are basically reinventing the accessors in `bzl.Rule`.
Why don't you wrap your `bzl.CallExpr` with `bzl.Rule` and reuse the accessors in `bzl.Rule`?

File: go/tools/gazelle/merger/merger.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+	for _, item := range src.List {
+		nkv, ok := item.(*bzl.BinaryExpr)
+		if !ok {
+			continue
+		}
+		k := literal(nkv.X)
+		if !mergeableFields[k] {
+			continue
+		}
+		// TODO(pmbethe09): allow '# keep' on src files.
```
Comment (yugui) [inline go/tools/gazelle/merger/merger.go]: 👍

File: go/tools/gazelle/merger/merger_test.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        "parse_test.go",
+        "print_test.go",
+    ],
+    data = glob(["testdata/*"]),
+    library = ":go_default_library",
+)
+`
+
+func TestMergeWithExisting(t *testing.T) {
+	tmp, err := ioutil.TempFile("", "")
```
Comment (yugui) [inline go/tools/gazelle/merger/merger_test.go]: This time, `os.Getenv("TEST_TMPDIR")` should work unlike before. 😉

File: go/tools/gazelle/merger/merger_test.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.Remove(tmp.Name())
+	if _, err := io.Copy(tmp, bytes.NewBufferString(oldData)); err != nil {
```
Comment (yugui) [inline go/tools/gazelle/merger/merger_test.go]: `io.WriteString`?

File: go/tools/gazelle/merger/merger.go (hunk 3 above)
Comment (pmbethe09) [inline go/tools/gazelle/merger/merger.go]: I have thought about this, but wanted to get the baseline in before working on it.
For example, in buildifier we have the generated file "parse.y.go" which will need a # keep before gazelle will work correctly.

File: go/tools/gazelle/merger/merger.go (hunk 2 above)
Comment (pmbethe09) [inline go/tools/gazelle/merger/merger.go]: I must have missed those when coding this up.  Let me fix up.

File: go/tools/gazelle/merger/merger_test.go (hunk 5 above)
Comment (pmbethe09) [inline go/tools/gazelle/merger/merger_test.go]: done

File: go/tools/gazelle/merger/merger.go (hunk 2 above)
Comment (pmbethe09) [inline go/tools/gazelle/merger/merger.go]: done, except stringValue, which I couldn't find a clean replacement for.

**Comment by yugui** [comment]
LGTM.

* Travis CI is failing because this #143 has not yet been merged. Let's ignore the failure for now.
* Jenkins build failed because of disk full in the instance.

Jenkins, retest this please.

**Pull request merged**